File: src/adapters/driver/http/handlers/allStreamsPlaylistHandler.go
package handlers

import (
	"log"
	"net/http"

	"Theatrum/domain/services"
)

type AllStreamsPlaylistHandler struct {
	applicationService *services.ApplicationService
	streamService     *services.StreamService
}

func NewAllStreamsPlaylistHandler(applicationService *services.ApplicationService, streamService *services.StreamService) *AllStreamsPlaylistHandler {
	return &AllStreamsPlaylistHandler{
		applicationService: applicationService,
		streamService:     streamService,
	}
}

func (h *AllStreamsPlaylistHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
	content, err := h.applicationService.BuildAllStreamsPlaylist()
	if err != nil {
		log.Printf("Error building all streams playlist: %v", err)
		http.Error(w, "Error building all streams playlist", http.StatusInternalServerError)
		return
	}

	// Set response headers
	w.Header().Set("Content-Type", "application/x-mpegURL")
	w.WriteHeader(http.StatusOK)

	// Write the response
	if _, err := w.Write([]byte(content)); err != nil {
		log.Printf("Error writing all streams playlist response: %v", err)
		http.Error(w, "Error encoding response", http.StatusInternalServerError)
		return
	}
} 

File: src/main/java/com/fjourdren/theatrum/infrastructure/adapter/in/web/handlers/AllStreamsPlaylistHandler.java
package com.fjourdren.theatrum.infrastructure.adapter.in.web.handlers;

import com.fjourdren.theatrum.application.port.in.ResolveChannelUseCase;
import com.fjourdren.theatrum.infrastructure.adapter.in.web.HttpConstants;
import jakarta.servlet.http.HttpServletResponse;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Component;

import java.io.IOException;

/** Serves the M3U8 playlist listing every available stream. */
@RequiredArgsConstructor
@Slf4j
@Component
public class AllStreamsPlaylistHandler {

    private final ResolveChannelUseCase applicationService;

    public void handle(HttpServletResponse response) throws IOException {
        String content;
        try {
            content = applicationService.buildAllStreamsPlaylist();
        } catch (RuntimeException e) {
            log.error("Error building all streams playlist: {}", e.getMessage());
            HttpErrors.send(response, HttpServletResponse.SC_INTERNAL_SERVER_ERROR, "Error building all streams playlist");
            return;
        }

        response.setContentType(HttpConstants.CONTENT_TYPE_M3U);
        response.setStatus(HttpServletResponse.SC_OK);
        response.getWriter().write(content);
    }
}
